methylpy/utilities.py

```diff
@@ -681,34 +681,46 @@ def index_allc_file(allc_file,reindex=False):
             pass
         else:
             return 0
-    g = open(index_file,'w')
-    f = open_allc_file(allc_file)
-    cur_chrom = ""
+
+    if allc_file.endswith('gz'):  # works for .gz, .bgz
+        f = subprocess.Popen(['zcat', allc_file],
+                             stdout=subprocess.PIPE,
+                             encoding='utf8').stdout
+    else:
+        f = open(allc_file)
+
+    index_lines = []
+    cur_chrom = "TOTALLY_NOT_A_CHROM"
+    cur_start = cur_chrom + '\t'
+    cur_pointer = 0
     # check header
-    line = f.readline()
-    try:
-        fields = line.split("\t")
-        int(fields[1])
-        int(fields[4])
-        int(fields[5])
-        # no header, continue to start from the beginning of allc file
-        f.seek(0)
-        cur_pointer = 0
-    except:
-        # find header, skip it
-        cur_pointer = f.tell()
-    # find chrom pointer
-    while True:
-        line = f.readline()
-        if not line: break
-        fields = line.split("\t")
-        if fields[0] != cur_chrom:
-            g.write(fields[0]+"\t"+str(cur_pointer)+"\n")
+    first_line = True
+    for line in f:
+        if first_line:
+            # all new allc files don't have header,
+            # some old files' first line is header
+            fields = line.split("\t")
+            first_line = False
+            try:
+                int(fields[1])
+                int(fields[4])
+                int(fields[5])
+                # no header, continue to start from the beginning of allc file
+            except ValueError:
+                # find header, skip it
+                cur_pointer += len(line)
```

Comment:
The meaning of `cur_pointer` is changed to the starting line number of the data of each chromosome from the starting byte of data of each chromosome. This will likely break methylpy. Can you double check?

Reply:
cur_pointer is not counting line number but count characters

Reply:
Sorry. Still counting characters is different from counting bytes. It probably will break methylpy. You may need to change the `open_allc_file` function as well.

Reply:
For example (https://github.com/yupenghe/methylpy/blob/4f634eeb26d718f14f4251c59e18b9d0a1bf9517/methylpy/utilities.py#L570), the index is used like this
```
fhandles[index].seek(chrom_pointer[index].get(chrom,0))
```
This will be a problem if the unit of `chrom_pointer[index].get(chrom,0)` is character instead of byte.

Reply:
I tested with your open_allc_file, you use 'rt' to open gzip, I think f.seek on the handle is seeking characters.

Here is a test:

```python
allc_path = './allc_180322_CEMBA_mm_P56_P63_1A_CEMBA180226_1A_1_CEMBA180226_1A_2_A10_AD001_indexed.tsv.gz'
```


```python
%time index_allc_file(allc_path)
```

    CPU times: user 11.2 s, sys: 303 ms, total: 11.5 s
    Wall time: 11.5 s

    0


```python
import gzip

# your open function
def open_allc_file(allc_file):
    if allc_file[-3:] == ".gz":
        f = gzip.open(allc_file,'rt')
    else:
        f = open(allc_file,'r')
    return(f)

def test_seek(allc_path, pointer):
    n = 0
    f = open_allc_file(allc_path)
    f.seek(pointer)
    for line in f:
        n += 1
        if n % 5 == 0:
            break
        print(line, end='')
    f.close()
```


```python
# seek chrM
test_seek(allc_path, 5980)
```

    chrM	238	-	CTG	0	1	1
    chrM	241	-	CTG	0	1	1
    chrM	243	-	CAC	0	1	1
    chrM	255	-	CTT	0	1	1



```python
# seek one line before
test_seek(allc_path, 5957)
```

    chrL	40240	+	CAG	0	1	1
    chrM	238	-	CTG	0	1	1
    chrM	241	-	CTG	0	1	1
    chrM	243	-	CAC	0	1	1



```python
# seek chr17
test_seek(allc_path, 1065690100)
```

    chr17	3003443	+	CAA	0	1	1
    chr17	3004010	-	CTA	0	1	1
    chr17	3004012	-	CTC	0	1	1
    chr17	3004014	-	CTC	0	1	1



```python
# seek one line before
test_seek(allc_path, 1065690073)
```

    chr16	98107534	+	CGG	1	1	1
    chr17	3003443	+	CAA	0	1	1
    chr17	3004010	-	CTA	0	1	1
    chr17	3004012	-	CTC	0	1	1

Reply:
idx file generated by this function:
```
chrL	0
chrM	5980
chrX	30754
chrY	32007248
chr1	34258004
chr2	122728442
chr3	208012827
chr4	279657470
chr5	352690628
chr6	423364987
chr7	492449464
chr8	558985055
chr9	619316679
chr10	679149881
chr11	740229483
chr12	804617978
chr13	859953362
chr14	916471801
chr15	970802307
chr16	1020817280
chr17	1065690100
chr18	1111475334
chr19	1153608335
#eof
```

Reply:
Ok. This looks great. Sorry I misunderstood the seek function.

Reply:
I also tested on unzipped ALLC, the seek position is the same.